src/tasks/withdrawService.ts

```diff
@@ -0,0 +1,385 @@
+import "@nomiclabs/hardhat-ethers";
+
+import { promises as fs } from "fs";
+
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { assert } from "chai";
+import { utils, Contract } from "ethers";
+import { task, types } from "hardhat/config";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+
+import { Api, Environment } from "../services/api";
+import { BUY_ETH_ADDRESS } from "../ts";
+
+import { dump, MAX_ORDER_VALIDITY_SECONDS } from "./dump";
+import {
+  getDeployedContract,
+  isSupportedNetwork,
+  SupportedNetwork,
+} from "./ts/deployment";
+import { sleep } from "./ts/sleep";
+import { balanceOf, erc20Token } from "./ts/tokens";
+import { ReferenceToken, REFERENCE_TOKEN } from "./ts/value";
+import { withdraw } from "./withdraw";
+import { getAllTradedTokens } from "./withdraw/traded_tokens";
+
+const MAX_ORDER_RETRIES_BEFORE_SKIPPING = 10;
+const MAX_CHECKED_TOKENS_PER_RUN = 200;
+
+export interface State {
+  /**
+   * Latest block number when the balances were withdrawn in the previous run.
+   */
+  lastUpdateBlock: number;
+  /**
+   * All tokens ever traded on GPv2. Stored in the state to avoid stressing the
+   * node by recovering the list every time.
+   */
+  tradedTokens: string[];
+  /**
+   * Index (in tradedTokens) following that of the last withdrawn token in the
+   * previous run of this script.
+   */
+  nextTokenToTrade: number;
+  /**
+   * Tokens that have a pending order from a previous execution of the withdraw
+   * service.
+   */
+  pendingTokens: PendingToken[];
+}
+
+interface PendingToken {
+  /**
+   * The address of the token.
+   */
+  address: string;
+  /**
+   * How many consecutive times the script sold this token with no success.
+   * Note that the script will not wait to check if the orders were successful,
+   * which means that every order in a run will add a pending token with one
+   * retry.
+   */
+  retries: number;
+}
+
+function isValidState(state: unknown): state is State {
```

Comment:
I'm not a typescript expert but whenever I see some quite big chunk of custom boilerplate code I ask myself "do we really need to implement this ourselves"? According to https://blog.logrocket.com/dynamic-type-validation-in-typescript/ there are some libraries which could validate a schema for us. Up to you if that is easier to use. My concern about this validation is that it wouldn't force us to update it if e.g. we add a field to the state enum (it would probably also require some unit tests for completeness).

Reply:
Implemented in #779.